usr/src/tools/scripts/acr.1
.\" ident	"@(#)acr.1	1.3	06/07/12 SMI"
.\" " CDDL HEADER START
.\" "
.\" " The contents of this file are subject to the terms of the
.\" " Common Development and Distribution License (the "License").
.\" " You may not use this file except in compliance with the License.
.\" "
.\" " You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" " or http://www.opensolaris.org/os/licensing.
.\" " See the License for the specific language governing permissions
.\" " and limitations under the License.
.\" "
.\" " When distributing Covered Code, include this CDDL HEADER in each
.\" " file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" " If applicable, add the following below this CDDL HEADER, with the
.\" " fields enclosed by brackets "[]" replaced with your own identifying
.\" " information: Portions Copyright [yyyy] [name of copyright owner]
.\" "
.\" " CDDL HEADER END
.\" "
.\" " Copyright 2006 Sun Microsystems, Inc.  All rights reserved.
.\" " Use is subject to license terms.
.TH acr 1 "07/13/06"
.SH NAME
.I acr
\- Automatically attempt BFU conflict resolution.
.SH SYNOPSIS
\fBacr [<root> [<archive-dir>]]\fP
.LP
.SH DESCRIPTION
.IX "OS-Net build tools" "acr" "" "\fBacr\fP"
.LP
.I acr 
may be used after 
.I bfu
reports that it has discovered conflicts (changes to editable files),
in order to automatically resolve these conflicts using the
same scripts which are used by regular packaging.
.LP
It unpacks and uses the contents of a special archive,
"conflict_resolution.gz" ,
constructed by
.I mkacr,
which is normally run whenver a
.I nightly
build creates archives.
.LP
If invoked with no arguments (the usual for an in-place BFU), it
looks at the archive directory used by the last BFU on this root.
.LP 
If given with the name of an alternate root directory as its first
argument, it resolves conflicts in that alternate root.
That directory can be either the root of a local zone, a mounted Live
Upgrade Boot Environment, or any other BFU'ed solaris install.
.LP
If given the name of an archive directory as its second argument, it
searches for the conflict resolution archive in that directory.
.LP 
When operating on /, acr also resolves conflicts for all installed
zones.
.SH LIMITATIONS
While this script is usually much more reliable than manual conflict
resolution, bugs in the packaging scripts it uses may still result in
brickification.   When in doubt, check its work.
.LP
Use extreme care when bfu'ing to bits older than currently installed.
.LP
If a system has multiple zones, it is necessary to invoke acr once
per zone with conflicts.
.LP
The class-action scripts are not extensively tested for build-to-build
updates.
.SH EXAMPLES
.LP
.B
# bfu /ws/on10-gate/archives/sparc/nightly
.br
 ...
.br
# acr
.br
 ...
.br
# acr /export/zones/a/root
.br
.SH BUGS
.LP
Positional arguments suck.
.LP
.SH SEE ALSO
.BR nightly "(1), " mkacr "(1), " cpio "(1), " lumount "(1)"


usr/src/tools/scripts/acr.sh
#!/bin/ksh -p
#
# CDDL HEADER START
#
# The contents of this file are subject to the terms of the
# Common Development and Distribution License (the "License").
# You may not use this file except in compliance with the License.
#
# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
# or http://www.opensolaris.org/os/licensing.
# See the License for the specific language governing permissions
# and limitations under the License.
#
# When distributing Covered Code, include this CDDL HEADER in each
# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
# If applicable, add the following below this CDDL HEADER, with the
# fields enclosed by brackets "[]" replaced with your own identifying
# information: Portions Copyright [yyyy] [name of copyright owner]
#
# CDDL HEADER END
#
#
# Copyright 2006 Sun Microsystems, Inc.  All rights reserved.
# Use is subject to license terms.
#
#ident	"@(#)acr.sh	1.4	06/07/12 SMI"

#
# usage: acr [root [archivedir]]
#
# examples:  acr
#	     acr /export/home/zone/root
#	     acr /export/home/zone/root ${CODEMGR_WS}/archives/sparc/nightly
#

if [ $# -gt 2 ] ; then
	print -u2 "usage:  $0 <root> <archivedir>"
	exit 1
fi

root=${1-/}
archivedir=$2
if [ -z "$archivedir" -o "$archivedir" = again ]; then
	if [ ! -s $root/etc/motd ]; then
		print -u2 "$root/etc/motd not found; this doesn't look like a" \
		    "valid root."
		exit 1
	fi
	archivedir=$(nawk '/^bfu.ed from / { print $3; exit }' $root/etc/motd)
fi

if [ ! -d "$archivedir" ]; then
    	print -u2 "Archive directory '$archivedir' not found."
	exit 1
fi

#
# temporary file scorecard:
#
# conflictscripts	list of scripts we need to run.
#
# installnew		list of editable files without class action scripts
#
# processedscript	basename of class action script edited to NOP
#			installf/removef's and use /tmp/bfubin
#
# allresults		log of all class action script output.
#
# thisresult		log of most recent class action script
#

tmpdir=$(mktemp -t -d acr.XXXXXX)

if [ -z "$tmpdir" ] ; then
    	print -u2 "mktemp failed to produce output; aborting."
	exit 1
fi

if [ ! -d $tmpdir ] ; then
    	print -u2 "mktemp failed to create a directory; aborting."
	exit 1
fi

conflictscripts=$tmpdir/conflictscripts
installnew=$tmpdir/installnew
allresults=$tmpdir/allresults
thisresult=$tmpdir/thisresult
processedscript=$tmpdir/processedscript

#
# This file is left over (on purpose) by BFU so that in a post-BFU environment
# we know which zones BFU processed.
#
local_zone_info_file=$root/.bfu_local_zone_info

get_cr_archive() {
	compressed_archive=$archivedir/conflict_resolution.gz
	if [ ! -s $compressed_archive ] ; then
		print -u2 "Failed to find conflict resolution information" \
		    "at $compressed_archive."
		return 1
	fi

	print -n "Getting ACR information from $archivedir... "

	gzip -d -c $compressed_archive | \
	    (cd $tmpdir; cpio -idmucB > /dev/null 2>&1) || return 1

	crdir=$tmpdir/conflict_resolution

	if [ ! -d $crdir ] ; then
		print -u2 "The conflict resolution archive is missing the" \
		    "conflict_resolution directory."
		return 1
	fi

	if [ ! -f $crdir/editable_file_db ] ; then
		print -u2 "The conflict resolution archive is missing the" \
			"editable file list."
		return 1
	fi
	print "ok"
	return 0
}

#
# If we're running after a BFU, some behaviors are different.
#
if [ -d /tmp/bfubin ] ; then
	bfu_alt_reality="true"
else
	bfu_alt_reality="false"
fi


acr_a_root() {
	typeset root
	typeset zone

	root=$1
	zone=$2

	print "ZONE $2 on $1" >> $allresults


	rm -f $conflictscripts
	rm -f $installnew

	#
	# The files that need to be processed are those that were stored by
	# bfu in bfu.conflicts.  Just process those that changed in the
	# distribution since the last bfu (those that are listed in the
	# $root/bfu.conflicts/NEW file).
	#
	if [ ! -d $root/bfu.conflicts ] ; then
		print -u2 "No BFU conflict information."
		return 1
	fi

	if [ ! -s $root/bfu.conflicts/NEW ] ; then
		print "No conflicts to resolve."
		return 0
	fi

	#
	# Some class-action scripts rely on being run in the order defined in
	# packages.  The $crdir/editable_file_db file contains the classes
	# in the correct order, so we preserve that order.
	#
	cat $crdir/editable_file_db | \
	while read filename script pkg pkginst isa mach unique pkgdef
	do
		grep "^$filename\$" $root/bfu.conflicts/NEW >> /dev/null || \
		    continue

		if [ "$mach" != "-" -a  $(uname -m) != "$mach" ] ; then
			continue
		fi

		print $filename $script $pkg $pkginst $isa $mach $unique \
		    $pkgdef >> $conflictscripts
	done

	if [ ! -s $conflictscripts ] ; then
		print "\nNo upgrade scripts were found for any of the" \
		    "conflicting files."
		return 1
	fi


	#
	# Look for files that are in the conflict list, but don't have
	# entries in the $conflictscripts file.  These have no
	# install scripts, so should just be copied. (The fact that such
	# files exist indicates a bug in bfu or possibly in the Solaris
	# packaging.  If these are really editable files, they should have
	# class action scripts.  If not, bfu shouldn't special-case them.)
	#
	cat $root/bfu.conflicts/NEW | while read filename ; do
		grep "$filename " $conflictscripts >> /dev/null
		if [ $? != 0 ] ; then
			print $filename >> $installnew
		fi
	done

	if [ -s $installnew ] ; then
		print "\n    The following files did not have conflict" \
		    "resolution scripts; this may"
		print "    indicate a bug in BFU.  The new versions will be" \
		    "installed.  The previous"
		print "    versions of these files can be found in" \
		    "$root/bfu.child:\n"
		cat $installnew | sed 's/^/        /'
		print
	fi

	UPDATE=yes
	BASEDIR=$root
	PKGSAV=/tmp
	PKG_INSTALL_ROOT=$root
	export UPDATE BASEDIR PKGSAV PKG_INSTALL_ROOT

	column_fmt='    %-35s  %-20s '

	printf "\n$column_fmt %s\n" "FILE" "ACTION" "STATUS"
	cat $conflictscripts | while read filename script pkg pkginst isa mach \
	    unique pkgdef ; do
		if [ "$script" = "upgrade_default" ] ; then
			msg=`printf "$column_fmt" \
			    $filename "upgrade_default_copy"`
			cp $root/bfu.conflicts/$filename $root/$filename
			if [ $? != 0 ]; then
				printf "$msg ok\n"
				continue
			else
				printf "$msg FAIL\n" 1>&2
				return 1
			fi
		fi

		if [ "$unique" = "c" ] ; then
			scriptloc=$crdir/$pkgdef/$script
		else
			scriptloc=$crdir/$pkgdef/$pkginst/$script
		fi
		msg=`printf "$column_fmt" $filename $script`

		#
		# If we are running in the post-BFU alternate reality, we need
		# to modify the class action script to work in that alternate
		# reality.  Otherwise, we merely need to "neuter" installf
		# and removef.  In any event, skip this one and go to the next
		# if the sed fails.
		#
		if [ $bfu_alt_reality = "true" ] ; then
			sed -e 's,^#!/bin/sh,#!/tmp/bfubin/sh,' \
			    -e 's,/usr/bin/,/tmp/bfubin/,g' \
			    -e 's,/usr/bin:,/tmp/bfubin:,' \
			    -e 's,installf,/tmp/bfubin/true,' \
			    -e 's,removef,/tmp/bfubin/true,' \
			    $scriptloc > $processedscript.$script
			error=$?
		else
			sed -e 's,installf,/usr/bin/true,' \
			    -e 's,removef,/usr/bin/true,' \
			    $scriptloc > $processedscript.$script
			error=$?
		fi

		if [ $error != 0 ] ; then
			printf "$msg FAIL (sed surgery->$error)\n" 1>&2
			continue
		fi

		chmod +x $processedscript.$script
		error=$?
		if [ $error != 0 ] ; then
			printf "$msg FAIL (chmod->$error)\n" 1>&2
			continue
		fi

		PKG=$pkg
		PKGINST=$pkg
		if [ $mach = "-" ] ; then
			ARCH=$isa
		else
			ARCH=$isa.$mach
		fi
		export PKG PKGINST ARCH

		print "PROCESSING $filename with $script" >> $allresults
		print $root/bfu.conflicts/$filename $root/$filename |
			$processedscript.$script > $thisresult 2>&1
		error=$?
		if [ $error != 0 ] ; then
			printf "$msg FAIL (exit $error)\n" 1>&2
			print -u2 "    Output of upgrade script:"
			sed 's/^/        /' < $thisresult >&2
		else
			printf "$msg ok\n"
		fi
		cat $thisresult >> $allresults
		print "RETURN CODE: $error" >> $allresults
	done

	if [ -s $installnew ] ; then
		cat $installnew | while read filename ; do

			msg=`printf "$column_fmt" $filename \
			    "cp from new version"`
			cp $root/bfu.conflicts/$filename $root/$filename
			error=$?
			if [ $error != 0 ] ; then
				printf "$msg FAIL (exit $error)\n" 1>&2
			else
				printf "$msg ok\n"
			fi

		done
	fi
}

#
# If we're post-BFU, then BFU should have left us a file listing
# which zones it processed.  If we're not post-BFU, just process
# all installed zones.
#
if [ $bfu_alt_reality = "false" ]; then
	zoneadm list -pi | nawk -F: '{
		if ($3 == "installed") {
			printf "%s %s\n", $2, $4
		}
	}' > $local_zone_info_file
fi

#
# To be terse, check whether there is any work to do at all; if not,
# just print one line and exit.
#
need_resolve=false
if [ -s $root/bfu.conflicts/NEW ]; then
	need_resolve=true
else
	if [ -s $local_zone_info_file ]; then
		cat $local_zone_info_file | while read zone zonepath; do
			if [ -s $zonepath/root/bfu.conflicts/NEW ] ; then
				need_resolve=true
			fi
		done
	fi
fi

if [ "$need_resolve" = "false" ]; then
	print "No conflicts to resolve."
	exit 0
fi

get_cr_archive || exit 1

printf "\nProcessing global zone:\t"
acr_a_root $root "global"

if [ $root != "/" ]; then
	printf "\nSkipping non-global zones (root is not /)"
else
	if [ -s $local_zone_info_file ]; then
		cat $local_zone_info_file | while read zone zonepath; do
			printf "\nProcessing zone $zone:\t"
			acr_a_root $zonepath/root $zone
		done
	fi
fi

echo
bootadm -a update_all 2>&1
print "Finished.  See $allresults for complete log."
